cache/save_skip.go

```diff
@@ -1,48 +1,100 @@
 package cache
 
 import (
-	"fmt"
 	"strings"
 )
 
-func (s *saver) canSkipSave(keyTemplate, evaluatedKey string, onlyCheckCacheKey bool) (canSkip bool, reason string) {
+type skipReason int
+
+const (
+	reasonKeyNotDynamic skipReason = iota
+	reasonNoRestore
+	reasonRestoreSameUniqueKey
+	reasonRestoreSameKeyNotUnique
+	reasonRestoreOtherKey
+	reasonNewArchiveChecksumMatch
+	reasonNewArchiveChecksumMismatch
+)
+
+func (r skipReason) String() string {
+	switch r {
+	case reasonKeyNotDynamic:
+		return "key_not_dynamic"
+	case reasonNoRestore:
+		return "no_restore_found"
+	case reasonRestoreSameUniqueKey:
+		return "restore_same_unique_key"
+	case reasonRestoreSameKeyNotUnique:
+		return "restore_same_key_not_unique"
+	case reasonRestoreOtherKey:
+		return "restore_other_key"
+	case reasonNewArchiveChecksumMatch:
+		return "new_archive_checksum_match"
+	case reasonNewArchiveChecksumMismatch:
+		return "new_archive_checksum_mismatch"
+	default:
+		return "unknown"
+	}
+}
+
+func (r skipReason) description() string {
+	switch r {
+	case reasonKeyNotDynamic:
+		return "key is not dynamic; the expectation is that the same key is used for saving different cache contents over and over"
+	case reasonNoRestore:
+		return "no cache was restored in the workflow, creating a new cache entry"
+	case reasonRestoreSameUniqueKey:
+		return "a cache with the same key was restored in the workflow, new cache would have the same content"
+	case reasonRestoreSameKeyNotUnique:
+		return "a cache with the same key was restored in the workflow, but contents might have changed since then"
+	case reasonRestoreOtherKey:
+		return "there was no cache restore in the workflow with this key"
```

Comment:
The key name and description for the `reasonRestoreOtherKey` does not say exactly the same thing. The description says that there was no cache restore for this key. But then the key name actually indicates that it was for other keys. So I would probably combine the two have 
```
there was no cache restore in the workflow with this key, but was for other(s)
```
or something like that. 

This came up as I was looking at this code piece:
```
s.logger.Infof("Can't skip saving the cache, reason: %s", reason.description())
    if reason == reasonRestoreOtherKey {
        s.logOtherHits()
 }
```
and I was wondering why do you print out the other hits when that reason means `there was no cache restore in the workflow with this key`. Then I read more of the code and understood it.

Reply:
Thank you for flagging this, it's indeed confusing. I improved the enum name and the description, please take a look